Updates to the vLLM optimization guide for MI300X/MI355X

peterjunpark opened the pull request:
## Motivation

Updates to vLLM workload optimization guide by @pinsiangamd @hongxiayang 


    Add AITER fused-kernel switches
    Introduce Quick Reduce for large all-reduces
    attention backends: Triton Unified, Triton Prefill-Decode, AITER MHA/MLA
    Update engine defaults (max-num-seqs, max-num-batched-tokens, async scheduling)
    Expand parallelism guidance (TP/PP/DP/EP)
    Modernize benchmarking section (multi-instance per-node)
    Add compilation section (graph modes, compile/capture sizes)
    Broaden quantization coverage (FP4/FP8, on-the-fly FP8, FP8 KV cache)
    Add speculative decoding example


<!-- Explain the purpose of this PR and the goals it aims to achieve. -->

## Technical Details

<!-- Explain the changes along with any relevant GitHub links. -->

## Test Plan

<!-- Explain any relevant testing done to verify this PR. -->

## Test Result

<!-- Briefly summarize test outcomes. -->

## Submission Checklist

- [ ] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

tanpinsiang committed:
Expand vLLM optimization guide for MI300X/MI355X with comprehensive AITER coverage. attention backend selection, environment variables (HIP/RCCL/Quick Reduce), parallelism strategies, quantization (FP8/FP4), engine tuning, CUDA graph modes, and multi-node scaling.

tanpinsiang committed:
fix spelling and .wordlist.txt

hongxiayang committed:
Update docs/how-to/rocm-for-ai/inference-optimization/workload.rst

Co-authored-by: peterjunpark <git@peterjunpark.com>

hongxiayang committed:
replace accelerator with GPU

tanpinsiang committed:
update examples command for deepseek on DP and EP and EP

tanpinsiang committed:
Merge branch 'develop' of github.com-amd:pinsiangamd/ROCm-internal into develop

hongxiayang committed:
add note about speculative decoding throughput loss on more speculative draft tokens

tanpinsiang committed:
added notes for DP and MXFP4

tanpinsiang committed:
Merge branch 'develop' of github.com-amd:pinsiangamd/ROCm-internal into develop

pinsiangamd committed:
Merge branch 'develop' into develop

hongxiayang committed:
Update docs/how-to/rocm-for-ai/inference-optimization/workload.rst

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
-:doc:`/how-to/rocm-for-ai/inference/benchmark-docker/vllm`.
+   When ``VLLM_ROCM_USE_AITER=1``, most AITER component flags (LINEAR, MOE, RMSNORM, MLA, MHA, FP8BMM) automatically default to **True**. You typically only need to set the master switch ``VLLM_ROCM_USE_AITER=1`` to enable all optimizations.
+   ROCm provides a prebuilt optimized Docker image for validating the performance
+   of LLM inference with vLLM on MI300X Series GPUs. The Docker image includes
```
Don't use MI300X series.

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
 
-vLLM is a high-throughput and memory efficient inference and serving engine for large language models that has gained traction in the AI community for
-its performance and ease of use. See :ref:`fine-tuning-llms-vllm` for a primer on vLLM with ROCm.
+This guide helps you maximize vLLM throughput and minimize latency on AMD MI300X/MI325X/MI355X GPUs. Learn how to:
```
```suggestion
This guide helps you maximize vLLM throughput and minimize latency on AMD Instinct MI300X/MI325X/MI355X GPUs. Learn how to:
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+* Choose **parallelism strategies** (tensor, pipeline, data, expert) for multi-GPU deployments
+* Apply **quantization** (FP8/FP4) to reduce memory usage by 2-4× with minimal accuracy loss
+* Tune **engine arguments** (batch size, memory utilization, graph modes) for your use case
+* Benchmark and scale across **single-node** and **multi-node** configurations
```
```suggestion
* Enable **AITER** (AI Tensor Engine for ROCm) for speedups on LLM models.
* Configure **environment variables** for optimal HIP, RCCL, and Quick Reduce performance.
* Select the right **attention backend** for your workload (AITER MHA/MLA vs Triton).
* Choose **parallelism strategies** (tensor, pipeline, data, expert) for multi-GPU deployments.
* Apply **quantization** (``FP8``/``FP4``) to reduce memory usage by 2-4× with minimal accuracy loss.
* Tune **engine arguments** (batch size, memory utilization, graph modes) for your use case.
* Benchmark and scale across **single-node** and **multi-node** configurations.
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+This guide helps you maximize vLLM throughput and minimize latency on AMD MI300X/MI325X/MI355X GPUs. Learn how to:
+
+* Enable **AITER** (AI Tensor Engine for ROCm) for speedups on LLM models
+* Configure **environment variables** for optimal HIP, RCCL, and Quick Reduce performance
```
Check is Quick Reduce capitalized

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
-The following performance tips are not *specific* to vLLM -- they are general
-but relevant in this context. You can tune the following vLLM parameters to
-achieve optimal request latency and throughput performance.
+The following variables are generally useful for MI300X/MI355X and vLLM:
```
```suggestion
The following variables are generally useful for Instinct MI300X/MI355X GPUs and vLLM:
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
-  <https://instinct.docs.amd.com/projects/amdgpu-docs/en/latest/system-optimization/mi300x.html#environment-variables>`_,
-  the environment variable ``HIP_FORCE_DEV_KERNARG`` can improve vLLM
-  performance. Set it to ``export HIP_FORCE_DEV_KERNARG=1``.
+* **HIP & math libraries**
```
```suggestion
* **HIP and math libraries**
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
 
-Auto-tuning using PyTorch TunableOp
-------------------------------------
+  * ``export NCCL_MIN_NCHANNELS=112`` — increases RCCL channels from default (typically 32-64) to 112 on MI300X. **Only beneficial for multi-GPU distributed workloads** (tensor parallelism, pipeline parallelism). Single-GPU inference does not need this.
```
```suggestion
  * ``export NCCL_MIN_NCHANNELS=112`` — increases RCCL channels from default (typically 32-64) to 112 on the Instinct MI300X. **Only beneficial for multi-GPU distributed workloads** (tensor parallelism, pipeline parallelism). Single-GPU inference does not need this.
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+------------------------------------------
 
-* Enable TunableOp and tuning. Optionally, enable verbose mode:
+**AITER** (AI Tensor Engine for ROCm) provides ROCm-specific fused kernels optimized for MI300X/MI355X GPUs in vLLM V1.
```
```suggestion
**AITER** (AI Tensor Engine for ROCm) provides ROCm-specific fused kernels optimized for Instinct MI300X/MI355X GPUs in vLLM V1.
```

amd-jnovotny replied:
Probably don't need the full spelling both here and in the title.

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+**AITER** (AI Tensor Engine for ROCm) provides ROCm-specific fused kernels optimized for MI300X/MI355X GPUs in vLLM V1.
 
-  .. code-block:: shell
+**How AITER flags work:**
```
```suggestion
How AITER flags work:
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+**How AITER flags work:**
 
-     PYTORCH_TUNABLEOP_ENABLED=1 PYTORCH_TUNABLEOP_VERBOSE=1 your_vllm_script.sh
+* ``VLLM_ROCM_USE_AITER`` is the **master switch** (defaults to **False/0**)
```
```suggestion
* ``VLLM_ROCM_USE_AITER`` is the master switch (defaults to ``False``/``0``)
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
 
-     PYTORCH_TUNABLEOP_ENABLED=1 PYTORCH_TUNABLEOP_VERBOSE=1 your_vllm_script.sh
+* ``VLLM_ROCM_USE_AITER`` is the **master switch** (defaults to **False/0**)
+* Individual feature flags (``VLLM_ROCM_USE_AITER_LINEAR``, ``VLLM_ROCM_USE_AITER_MOE``, etc.) default to **True** but only activate when the master switch is enabled
```
```suggestion
* Individual feature flags (``VLLM_ROCM_USE_AITER_LINEAR``, ``VLLM_ROCM_USE_AITER_MOE``, etc.) default to ``True`` but only activate when the master switch is enabled
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+* To enable a specific AITER feature, you must set **both** ``VLLM_ROCM_USE_AITER=1`` **and** the specific feature flag to ``1``
 
-* Enable TunableOp and disable tuning and measure.
+**Quick start examples:**
```
```suggestion
Quick start examples:
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
 
-*  ``gpu_memory_utilization``
+   * - ``VLLM_ROCM_USE_AITER``
+     - **Master switch** to enable AITER kernels (**0/False** by default). All other ``VLLM_ROCM_USE_AITER_*`` flags require this to be set to ``1``.
```
```suggestion
     - **Master switch** to enable AITER kernels (``0``/``False`` by default). All other ``VLLM_ROCM_USE_AITER_*`` flags require this to be set to ``1``.
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
 
-*  ``dtype``
+   * - ``VLLM_ROCM_USE_AITER_LINEAR``
+     - Use AITER quantization operators + GEMM for linear layers (defaults to **True** when AITER is on). Accelerates matrix multiplications in all transformer layers. **Recommended: keep enabled.**
```
```suggestion
     - Use AITER quantization operators + GEMM for linear layers (defaults to ``True`` when AITER is on). Accelerates matrix multiplications in all transformer layers. **Recommended: keep enabled.**
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
 
-*  ``enforce_eager``
+   * - ``VLLM_ROCM_USE_AITER_MOE``
+     - Use AITER fused-MoE kernels (defaults to **True** when AITER is on). Accelerates Mixture-of-Experts routing and computation. **See detailed requirements below.**
```
I won't keep highlighting True from bold to code ticks. But they are throughout this section.

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
 
-*  ``input_len``
+   * - ``VLLM_ROCM_USE_AITER_MLA``
+     - Use AITER Multi-head Latent Attention for supported models e.g. DeepSeek-V3/R1 (defaults to **True** when AITER is on). **See detailed requirements below.**
```
```suggestion
     - Use AITER Multi-head Latent Attention for supported models, for example, DeepSeek-V3/R1 (defaults to **True** when AITER is on). **See detailed requirements below.**
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
 
-*  ``output_len``
+   * - ``VLLM_ROCM_USE_AITER_MHA``
+     - Use AITER Multi-Head Attention kernels (defaults to **True** when AITER is on; set **0** to use Triton attention backends and Prefill-Decode attention backend instead). **See attention backend selection below.**
```
```suggestion
     - Use AITER Multi-Head Attention kernels (defaults to ``True`` when AITER is on; set to ``0`` to use Triton attention backends and Prefill-Decode attention backend instead). **See attention backend selection below.**
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
 
-*  ``max_num_seqs``
+   * - ``VLLM_ROCM_USE_AITER_UNIFIED_ATTENTION``
+     - Enable AITER's optimized unified attention kernel (defaults to **False**). Only takes effect when: (1) AITER is enabled, (2) unified attention mode is active (``VLLM_V1_USE_PREFILL_DECODE_ATTENTION=0``), and (3) AITER MHA is disabled (``VLLM_ROCM_USE_AITER_MHA=0``). When disabled, falls back to vLLM's Triton unified attention.
```
```suggestion
     - Enable AITER's optimized unified attention kernel (defaults to **False**). Only takes effect when: AITER is enabled; unified attention mode is active (``VLLM_V1_USE_PREFILL_DECODE_ATTENTION=0``); and AITER MHA is disabled (``VLLM_ROCM_USE_AITER_MHA=0``). When disabled, falls back to vLLM's Triton unified attention.
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
 
-*  ``num_scheduler_steps``
+   * - ``VLLM_ROCM_USE_AITER_FP8BMM``
+     - Use AITER FP8 batched matmul (defaults to **True** when AITER is on). Fuses FP8 per-token quantization with batched GEMM (used in MLA models like DeepSeek-V3). **Requires MI300X/MI355X.**
```
```suggestion
     - Use AITER ``FP8`` batched matmul (defaults to **True** when AITER is on). Fuses ``FP8`` per-token quantization with batched GEMM (used in MLA models like DeepSeek-V3). **Requires an Instinct MI300X/MI355X GPU.**
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
 
-*  ``enable_chunked_prefill``
+   * - ``VLLM_ROCM_FP8_PADDING``
+     - Pad FP8 linear weight tensors to improve memory locality (defaults to **True**). Minor memory overhead for better performance.
```
```suggestion
     - Pad ``FP8`` linear weight tensors to improve memory locality (defaults to **True**). Minor memory overhead for better performance.
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
 
-*  ``distributed_executor_backend``
+   * - ``VLLM_ROCM_MOE_PADDING``
+     - Pad MoE weight tensors for better memory access patterns (defaults to **True**). Same memory/performance tradeoff as FP8 padding.
```
```suggestion
     - Pad MoE weight tensors for better memory access patterns (defaults to **True**). Same memory/performance tradeoff as ``FP8`` padding.
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
-  ``ShareGPT_V3_unfiltered_cleaned_split.json``. This dataset includes real-world conversational
-  data, making it a good representation of typical use cases for language models. Download it using
-  the following command:
+**When to enable:**
```
```suggestion
When to enable:
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+**When to enable:**
 
-  .. code-block:: shell
+* **Enable (default):** For all MoE models on MI300X/MI355X for best throughput
```
```suggestion
* **Enable (default):** For all MoE models on the Instinct MI300X/MI355X for best throughput
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+* **Disable:** Only for debugging or if you encounter numerical issues
 
-     wget https://huggingface.co/datasets/anon8231489123/ShareGPT_Vicuna_unfiltered/resolve/main/ShareGPT_V3_unfiltered_cleaned_split.json
+**Example usage:**
```
```suggestion
Example usage:
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
-For models that fit in one GPU -- to maximize the accumulated throughput -- you can run as many as eight vLLM instances
-simultaneously on one MI300X node (with eight GPUs). To do so, use the GPU isolation environment
-variable ``CUDA_VISIBLE_DEVICES``.
+**Critical requirement:**
```
```suggestion
Critical requirement:
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
-The total throughput achieved by running ``N`` instances of vLLM is generally much higher than running a
-single vLLM instance across ``N`` GPUs simultaneously (that is, configuring ``tensor_parallel_size`` as N or
-using the ``-tp`` N option, where ``1 < N ≤ 8``).
+**Applicable models:**
```
```suggestion
Applicable models:
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+* Other models using multi-head latent attention (MLA) architecture
 
-.. _mi300x-vllm-gpu-memory-utilization:
+**Example usage:**
```
```suggestion
Example usage:
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+Understanding which attention backend to use helps optimize your deployment.
 
-   .. code-block:: shell
+**Quick Reference: Which Attention Backend Will I Get?**
```
```suggestion
Quick reference: Which attention backend will I get?
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+**Quick Reference: Which Attention Backend Will I Get?**
 
-      /vllm-workspace/benchmarks/benchmark_throughput.py --gpu-memory-utilization 0.98 --input-len 1024 --output-len 128 --model /path/to/model
+**Default Behavior (No Configuration)**
```
```suggestion
Default behavior (no configuration)
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
 
-   For example, run two instances of the Llama3-8b model at the same time on a single GPU by specifying
-   ``--gpu-memory-utilization`` to 0.4 (40%) as follows (on GPU ``0``):
+**Recommended: Enable AITER (Set VLLM_ROCM_USE_AITER=1)**
```
```suggestion
**Recommended: Enable AITER (set ``VLLM_ROCM_USE_AITER=1``)**
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+   │
+   └─ NO  → AITER MHA Backend
+             • For standard transformer models (Llama, Mistral, etc.)
+             • Optimized for MI300X/MI355X
```
```suggestion
             • Optimized for Instinct MI300X/MI355X
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+             • Automatically selected
 
-See :ref:`vllm-engine-args` for other performance suggestions.
+**Advanced: Manual Backend Selection**
```
```suggestion
Advanced: Manual backend selection
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
 
-The two main reasons to use multiple GPUs are:
+   * - To Use This Backend
+     - Set These Flags
```
```suggestion
   * - To use this backend
     - Set these flags
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
 
-   python -m vllm.entrypoints.api_server --model /path/to/model --dtype
-   float16 -tp 2 --port 8000 &
+**Quick Start Examples:**
```
```suggestion
Quick start examples:
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
 
-Choose an attention backend
----------------------------
+**Which Backend Should I Choose?**
```
```suggestion
Which backend should I choose?
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
-  least once as a warm-up step so Triton can perform auto-tuning before
-  collecting benchmarking numbers. This is the default setting.
+   * - Your Use Case
+     - Recommended Backend
```
```suggestion
   * - Your use case
     - Recommended backend
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
-- **Composable Kernel (CK) Flash Attention** - To use CK Flash Attention, specify
-  the environment variable as ``export VLLM_USE_TRITON_FLASH_ATTN=0``.
+   * - **Standard transformer models** (Llama, Mistral, Qwen, Mixtral)
+     - **AITER MHA** (``VLLM_ROCM_USE_AITER=1``) — **Recommended for most workloads** on MI300X/MI355X. Provides optimized attention kernels for both prefill and decode phases.
```
```suggestion
     - **AITER MHA** (``VLLM_ROCM_USE_AITER=1``) — **Recommended for most workloads** on Instinct MI300X/MI355X. Provides optimized attention kernels for both prefill and decode phases.
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
 
-vLLM engine arguments
----------------------
+**Important Notes:**
```
```suggestion
**Important notes:**
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
 
-Use the float16 dtype
-^^^^^^^^^^^^^^^^^^^^^
+**How to Verify Which Backend is Active**
```
```suggestion
**How to verify which backend is active**
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+   * Generic implementation that works across all ROCm platforms
+   * Good baseline performance
+   * Automatically selected when ``VLLM_ROCM_USE_AITER=0`` (or unset)
+   * Support GPT-OSS
```
```suggestion
   * Supports GPT-OSS
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+
+2. **AITER Triton Unified Attention** (advanced, requires manual configuration)
+
+   * AMD's optimized unified Triton kernel
```
```suggestion
   * The AMD optimized unified Triton kernel
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+   * AMD's optimized unified Triton kernel
+   * Enable with ``VLLM_ROCM_USE_AITER=1``, ``VLLM_ROCM_USE_AITER_MHA=0``, and ``VLLM_ROCM_USE_AITER_UNIFIED_ATTENTION=1``
+   * Only useful for specific workloads; most users should use AITER MHA instead
+   * Recommended this backend when running GPT-OSS.
```
```suggestion
   * Enable with ``VLLM_ROCM_USE_AITER=1``, ``VLLM_ROCM_USE_AITER_MHA=0``, and ``VLLM_ROCM_USE_AITER_UNIFIED_ATTENTION=1``.
   * Only useful for specific workloads. Most users should use AITER MHA instead.
   * Recommended this backend when running GPT-OSS.
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
-Another vLLM performance tip is to enable chunked prefill to improve
-throughput. Chunked prefill allows large prefills to be chunked into
-smaller chunks and batched together with decode requests.
+3. **AITER Triton Prefill–Decode Attention** (hybrid, MI300X-optimized)
```
```suggestion
3. **AITER Triton Prefill–Decode Attention** (hybrid, Instinct MI300X-optimized)
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
-You can enable the feature by specifying ``--enable-chunked-prefill`` in the
-command line or setting ``enable_chunked_prefill=True`` in the LLM
-constructor. 
+   * Enable with ``VLLM_ROCM_USE_AITER=1`` and ``VLLM_ROCM_USE_AITER_MHA=0`` and ``VLLM_V1_USE_PREFILL_DECODE_ATTENTION=1``
```
```suggestion
   * Enable with ``VLLM_ROCM_USE_AITER=1``, ``VLLM_ROCM_USE_AITER_MHA=0``, and ``VLLM_V1_USE_PREFILL_DECODE_ATTENTION=1``
```

peterjunpark committed:
Apply suggestions from code review

Co-authored-by: Jeffrey Novotny <jnovotny@amd.com>

Update in-line formatting

Co-authored-by: Jeffrey Novotny <jnovotny@amd.com>

Bold sparingly

Co-authored-by: Jeffrey Novotny <jnovotny@amd.com>

formatting

Update docs/how-to/rocm-for-ai/inference-optimization/workload.rst

Co-authored-by: Jeffrey Novotny <jnovotny@amd.com>

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
-token) as you can put more prefill to the batch.
+     * **Prefill:** ``context_attention_fwd`` Triton kernel
+     * **Primary decode:** ``torch.ops._rocm_C.paged_attention`` (custom ROCm kernel optimized for head sizes 64/128, block sizes 16/32, GQA 1–16, context ≤131k; sliding window not supported)
+     * **Fallback decode:** ``kernel_paged_attention_2d`` Triton kernel when shapes don't meet primary decode requirements
```
```suggestion
     * **Prefill**: ``context_attention_fwd`` Triton kernel
     * **Primary decode**: ``torch.ops._rocm_C.paged_attention`` (custom ROCm kernel optimized for head sizes 64/128, block sizes 16/32, GQA 1–16, context ≤131k; sliding window not supported)
     * **Fallback decode**: ``kernel_paged_attention_2d`` Triton kernel when shapes don't meet primary decode requirements
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
-parameter ``max_num_batched_tokens`` (for example, increasing ``max_num_batched_tokens`` value to 4096 or larger).
+6. **AITER Multi-head Latent Attention (MLA)** (for DeepSeek-V3/R1/V2)
+
+   * Controlled by ``VLLM_ROCM_USE_AITER_MLA`` (**1** = enabled)
```
```suggestion
   * Controlled by ``VLLM_ROCM_USE_AITER_MLA`` (``1`` = enabled)
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+-----------------------------------------
+
+**Quick Reduce** is an alternative to RCCL/custom all-reduce for **large** inputs (MI300-class GPUs).
+It supports FP16/BF16 as well as symmetric INT8/INT6/INT4 quantized all-reduce (group size 32).
```
Types s/b in code ticks

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+* ``VLLM_ROCM_QUICK_REDUCE_CAST_BF16_TO_FP16``: cast BF16 input to FP16 (``1/True`` by default for performance).
+* ``VLLM_ROCM_QUICK_REDUCE_MAX_SIZE_BYTES_MB``: cap the preset buffer (default ``NONE`` ≈ ``2048`` MB).
+
+Quick Reduce tends to help **throughput** at higher TP counts (e.g., 4–8) with many concurrent requests.
```
```suggestion
Quick Reduce tends to help **throughput** at higher TP counts (for example, 4–8) with many concurrent requests.
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+
+**Choosing the right strategy:**
+
+* **Tensor Parallelism (TP)**: Use when model doesn't fit on one GPU. Prefer staying within a single XGMI island (≤8 GPUs on MI300X).
```
```suggestion
* **Tensor Parallelism (TP)**: Use when model doesn't fit on one GPU. Prefer staying within a single XGMI island (≤8 GPUs on the Instinct MI300X).
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+       --pipeline-parallel-size 2
+
+.. note::
+   **ROCm best practice:** On MI300X, prefer staying within a single XGMI island (≤8 GPUs) using TP only. Use PP when scaling beyond 8 GPUs or across nodes.
```
```suggestion
   **ROCm best practice**: On the Instinct MI300X, prefer staying within a single XGMI island (≤8 GPUs) using TP only. Use PP when scaling beyond eight GPUs or across nodes.
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+
+* Model fits on one GPU, but you need higher request throughput
+* Scaling across multiple nodes horizontally
+* Combining with tensor parallelism (e.g., DP=2 + TP=4 = 8 GPUs total)
```
```suggestion
* Combining with tensor parallelism (for example, DP=2 + TP=4 = 8 GPUs total)
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+
+**Performance considerations:**
+
+Expert parallelism is designed primarily for cross-node MoE deployments where high-bandwidth interconnects (like InfiniBand) between nodes make EP communication efficient. For single-node MI300X/MI355X deployments with XGMI connectivity, **tensor parallelism typically provides better performance** due to optimized all-to-all collectives on XGMI.
```
```suggestion
Expert parallelism is designed primarily for cross-node MoE deployments where high-bandwidth interconnects (like InfiniBand) between nodes make EP communication efficient. For single-node Instinct MI300X/MI355X deployments with XGMI connectivity, tensor parallelism typically provides better performance due to optimized all-to-all collectives on XGMI.
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+* Models with very large numbers of experts that benefit from expert distribution
+* Workloads where EP's reduced data movement outweighs communication overhead
+
+**Single-node recommendation:** For MI300X/MI355X within a single node (≤8 GPUs), prefer tensor parallelism over expert parallelism for MoE models to leverage XGMI's high bandwidth and low latency.
```
```suggestion
**Single-node recommendation:** For Instinct MI300X/MI355X within a single node (≤8 GPUs), prefer tensor parallelism over expert parallelism for MoE models to leverage XGMI's high bandwidth and low latency.
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+
+**Combining with Data Parallelism:**
+
+EP works seamlessly with Data Parallel Attention for optimal memory efficiency in MLA+MoE models (e.g., DeepSeek V3):
```
```suggestion
EP works seamlessly with Data Parallel Attention for optimal memory efficiency in MLA+MoE models (for example, DeepSeek V3):
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+         --dataset /path/to/ShareGPT_V3_unfiltered_cleaned_split.json &
+      done
+
+Total throughput from **N** single-GPU instances usually exceeds one instance stretched across **N** GPUs (`-tp N`).
```
```suggestion
Total throughput from **N** single-GPU instances usually exceeds one instance stretched across **N** GPUs (``-tp N``).
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+
+Total throughput from **N** single-GPU instances usually exceeds one instance stretched across **N** GPUs (`-tp N`).
+
+**Model coverage**: Llama 2 (7B/13B/70B), Llama 3 (8B/70B), Qwen2 (7B/72B), Mixtral-8x7B/8x22B, etc. Llama2‑70B
```
```suggestion
**Model coverage**: Llama 2 (7B/13B/70B), Llama 3 (8B/70B), Qwen2 (7B/72B), Mixtral-8x7B/8x22B, and others Llama2‑70B
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
-``fp8(w8a8)`` and ``AWQ`` quantization are supported for ROCm.
+Selected arguments that often help on ROCm. See `engine args docs <https://docs.vllm.ai/en/latest/serving/engine_args.html>`_ for the full list.
+
+Configure ``--max-num-seqs``
```
```suggestion
Configure --max-num-seqs
```

amd-jnovotny replied:
I don't think we typically use code ticks in titles

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+   vllm serve <model> --max-num-seqs 128 --max-model-len 8192
 
-AWQ quantization
+Configure ``--max-num-batched-tokens``
```
```suggestion
Configure --max-num-batched-tokens
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+* Lower values improve **ITL** (less prefill interrupting decode).
+* Higher values improve **TTFT** (more prefill per batch).
+
+Defaults: **8192** for online serving, **16384** for offline. However, optimal values vary significantly by model size—smaller models can efficiently handle larger batch sizes. Setting it near ``--max-model-len`` mimics V0 behavior and often maximizes throughput.
```
```suggestion
Defaults: **8192** for online serving, **16384** for offline. However, optimal values vary significantly by model size. Smaller models can efficiently handle larger batch sizes. Setting it near ``--max-model-len`` mimics V0 behavior and often maximizes throughput.
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+
+2. Pre-quantized model through Quark and llm-compressor 
+
+Supported Quantization Methods
```
```suggestion
Supported quantization methods
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+
+1. On-the-fly quantization 
+
+2. Pre-quantized model through Quark and llm-compressor 
```
```suggestion
* On-the-fly quantization 

* Pre-quantized model through Quark and llm-compressor 
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+Supported Quantization Methods
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+vLLM on ROCm supports the following quantization methods for AMD MI300 series and MI355X GPUs:
```
```suggestion
vLLM on ROCm supports the following quantization methods for the AMD Instinct MI300 series and Instinct MI355X GPUs:
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+     - Precision
+     - ROCm Support
+     - Memory Reduction
+     - Best Use Case
```
```suggestion
   * - Method
     - Precision
     - ROCm support
     - Memory reduction
     - Best use case
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+     - Best Use Case
+   * - **FP8** (W8A8)
+     - 8-bit float
+     - ✅ Excellent
```
We probably don't need the check marks.

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+     - 8-bit float
+     - ✅ Excellent
+     - 2× (50%)
+     - High throughput, better than FP8
```
```suggestion
     - High throughput, better than ``FP8``
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+     - KV cache: 50%
+     - All inference workloads
+   * - **Quark (AMD)**
+     - FP8/MXFP4
```
```suggestion
     - ``FP8``/``MXFP4``
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+     - 2-4× (50-75%)
+     - AMD pre-quantized models
+   * - **compressed-tensors**
+     - W8A8 INT8/FP8
```
```suggestion
     - W8A8 ``INT8``/``FP8``
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+**Key:**
+
+- ✅ Excellent: Fully supported with optimized kernels
+- ✅ Good: Supported, may not have AMD-optimized kernels
```
```suggestion
- ✅ Good: Supported, might not have AMD-optimized kernels
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+
+AMD provides pre-quantized models optimized for ROCm. These models are ready to use with vLLM.
+
+**AMD Quark Quantized Models:**
```
```suggestion
**AMD Quark Quantized models**:
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+  * `Llama-3.1-405B-Instruct-MXFP4-Preview <https://huggingface.co/amd/Llama-3.1-405B-Instruct-MXFP4-Preview>`__ (MXFP4 for MI350/MI355)
+  * `DeepSeek-R1-0528-MXFP4-Preview <https://huggingface.co/amd/DeepSeek-R1-0528-MXFP4-Preview>`__ (MXFP4 for MI350/MI355)
+
+**Quick Start:**
```
```suggestion
**Quick start**:
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+      --dtype auto \
+      --tensor-parallel-size 1
+
+**Other Pre-quantized Models:**
```
```suggestion
**Other pre-quantized models**:
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+- **GPTQ models**: `Hugging Face gptq flag <https://huggingface.co/models?other=gptq>`_
+- **LLM Compressor models**: `Hugging Face compressed-tensors flag <https://huggingface.co/models?other=compressed-tensors>`_
+
+On-the-fly Quantization
```
```suggestion
On-the-fly quantization
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+On-the-fly Quantization
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+For models without pre-quantization, vLLM can quantize FP16/BF16 models at server startup.
```
```suggestion
For models without pre-quantization, vLLM can quantize ``FP16``/``BF16`` models at server startup.
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+
+For models without pre-quantization, vLLM can quantize FP16/BF16 models at server startup.
+
+**Supported Methods:**
```
```suggestion
**Supported methods**:
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+
+**Supported Methods:**
+
+- ``fp8``: Per-tensor FP8 weight and activation quantization
```
```suggestion
- ``fp8``: Per-tensor ``FP8`` weight and activation quantization
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+**Supported Methods:**
+
+- ``fp8``: Per-tensor FP8 weight and activation quantization
+- ``ptpc_fp8``: Per-token-activation per-channel-weight FP8 (better accuracy same FP8 speed). See `PTPC-FP8 on ROCm blog post <https://blog.vllm.ai/2025/02/24/ptpc-fp8-rocm.html>`_ for details
```
```suggestion
- ``ptpc_fp8``: Per-token-activation per-channel-weight ``FP8`` (better accuracy same ``FP8`` speed). See `PTPC-FP8 on ROCm blog post <https://blog.vllm.ai/2025/02/24/ptpc-fp8-rocm.html>`_ for details
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+      --dtype auto \
+      --tensor-parallel-size 4
+
+**Note:** On-the-fly quantization adds 2-5 minutes startup time but eliminates pre-quantization. For production with frequent restarts, use pre-quantized models.
```
```suggestion
**Note**: On-the-fly quantization adds two to five minutes of startup time but eliminates pre-quantization. For production with frequent restarts, use pre-quantized models.
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+**Recommendation:** For AMD MI300X, **AWQ with Triton kernels** or **FP8 quantization** may provide better
+performance due to ROCm-specific optimizations, but GPTQ is a viable alternative.
+
+**Using Pre-quantized GPTQ Models:**
```
```suggestion
**ROCm support status**:

- ✅ **Fully supported** - GPTQ kernels compile and run on ROCm via HIP
- ✅ **Pre-quantized models work** with standard GPTQ kernels

**Recommendation**: For the AMD Instinct MI300X, **AWQ with Triton kernels** or **FP8 quantization** might provide better
performance due to ROCm-specific optimizations, but GPTQ is a viable alternative.

**Using pre-quantized GPTQ models**:
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+      --dtype auto \
+      --tensor-parallel-size 1
+
+**Important Notes:**
```
```suggestion
Important notes:
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+**Important Notes:**
+
+- **Kernel support:** GPTQ uses standard HIP-compiled kernels on ROCm
+- **Performance:** AWQ with Triton kernels may offer better throughput on AMD GPUs due to ROCm optimizations
```
```suggestion
- **Performance:** AWQ with Triton kernels might offer better throughput on AMD GPUs due to ROCm optimizations
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+AWQ (Activation-aware Weight Quantization) is a 4-bit weight quantization technique that provides excellent
+model compression with minimal accuracy loss (<1%). ROCm supports AWQ quantization on AMD MI300 series and
```
```suggestion
model compression with minimal accuracy loss (<1%). ROCm supports AWQ quantization on the AMD Instinct MI300 series and
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+model compression with minimal accuracy loss (<1%). ROCm supports AWQ quantization on AMD MI300 series and
+MI355X GPUs with vLLM.
+
+**Using Pre-quantized AWQ Models:**
```
```suggestion
**Using pre-quantized AWQ models:**
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+* **dtype parameter:** AWQ requires ``--dtype auto`` or ``--dtype float16``. The ``--dtype`` flag controls
+  the **activation dtype** (FP16/BF16 for computations), not the weight dtype. AWQ weights remain as INT4
+  (4-bit integers) as specified in the model's quantization config, but are dequantized to FP16/BF16 during
+  matrix multiplication operations.
```
```suggestion
* **dtype parameter:** AWQ requires ``--dtype auto`` or ``--dtype float16``. The ``--dtype`` flag controls
  the **activation dtype** (``FP16``/``BF16`` for computations), not the weight dtype. AWQ weights remain as INT4
  (4-bit integers) as specified in the model's quantization config, but are dequantized to ``FP16``/``BF16`` during
  matrix multiplication operations.
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+* **Group size:** 128 is recommended for optimal performance/accuracy balance
+* **Model compatibility:** AWQ is primarily tested on Llama, Mistral, and Qwen model families
+
+Quark (AMD Quantization Toolkit)
```
```suggestion
Quark (AMD quantization toolkit)
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+Quark (AMD Quantization Toolkit)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+AMD Quark is AMD's quantization toolkit optimized for ROCm. It supports FP8 W8A8, MXFP4, W8A8 INT8, and
```
```suggestion
AMD Quark is the AMD quantization toolkit optimized for ROCm. It supports ``FP8 W8A8``, ``MXFP4``, ``W8A8 INT8``, and
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+
+AMD Quark is AMD's quantization toolkit optimized for ROCm. It supports FP8 W8A8, MXFP4, W8A8 INT8, and
+other quantization formats with native vLLM integration. The quantization format will automatically be inferred
+from the model config file, thus we can omit `--quantization quark`.
```
```suggestion
from the model config file, so you can omit ``--quantization quark``.
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+- **FP8 models**: ~50% memory reduction, 2× compression
+- **MXFP4 models**: ~75% memory reduction, 4× compression
+- **Embedded scales**: Quark FP8-KV models include pre-calibrated KV-cache scales
+- **Hardware optimized**: Leverages AMD MI300 series FP8 acceleration
```
```suggestion
- **Hardware optimized**: Leverages the AMD Instinct MI300 series ``FP8`` acceleration
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+
+For creating your own Quark-quantized models, see `Quark Documentation <https://quark.docs.amd.com/latest/>`_.
+
+FP8 `kv-cache` dtype
```
```suggestion
FP8 kv-cache dtype
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+
+FP8 KV-cache quantization reduces memory footprint by approximately 50%, enabling longer context lengths
+or higher concurrency. ROCm supports FP8 KV-cache with both ``fp8_e4m3`` and ``fp8_e5m2`` formats on
+AMD MI300 series and other CDNA™ GPUs.
```
```suggestion
AMD Instinct MI300 series and other CDNA™ GPUs.
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+or higher concurrency. ROCm supports FP8 KV-cache with both ``fp8_e4m3`` and ``fp8_e5m2`` formats on
+AMD MI300 series and other CDNA™ GPUs.
+
+Use ``--kv-cache-dtype fp8`` to enable FP8 KV-cache quantization. For best accuracy, use calibrated
```
```suggestion
Use ``--kv-cache-dtype fp8`` to enable ``FP8`` KV-cache quantization. For best accuracy, use calibrated
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+accuracy impact.
+
+
+**Quick Start (Dynamic Scaling):**
```
```suggestion
**Quick start (dynamic scaling)**:
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+      --calculate-kv-scales \
+      --gpu-memory-utilization 0.90
+
+**Calibrated Scaling (Advanced):**
```
```suggestion
**Calibrated scaling (advanced)**:
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+For optimal accuracy, pre-calibrate KV-cache scales using representative data. The calibration process:
+
+#. Runs the model on calibration data (512+ samples recommended)
+#. Computes optimal FP8 quantization scales for key/value cache tensors
```
```suggestion
#. Computes optimal ``FP8`` quantization scales for key/value cache tensors
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+#. Embeds these scales into the saved model as additional parameters
+#. vLLM loads the model and uses the embedded scales automatically when ``--kv-cache-dtype fp8`` is specified
+
+The quantized model can be used like any other model—the embedded scales are stored as part of the model weights.
```
```suggestion
The quantized model can be used like any other model. The embedded scales are stored as part of the model weights.
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+
+**Using pre-calibrated models:**
+
+AMD provides ready-to-use models with pre-calibrated FP8 KV cache scales:
```
```suggestion
AMD provides ready-to-use models with pre-calibrated ``FP8`` KV cache scales:
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+
+For detailed instructions and the complete calibration script, see the `FP8 KV Cache Quantization Guide <https://github.com/vllm-project/llm-compressor/blob/main/examples/quantization_kv_cache/README.md>`_.
+
+**Format Options:**
```
```suggestion
**Format options**:
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+Speculative decoding (experimental)
+-----------------------------------
+
+Recent vLLM versions add support for speculative decoding backends (e.g., Eagle‑v3). Evaluate for your model and latency/throughput goals.
```
```suggestion
Recent vLLM versions add support for speculative decoding backends (for example, Eagle‑v3). Evaluate for your model and latency/throughput goals.
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+
+Recent vLLM versions add support for speculative decoding backends (e.g., Eagle‑v3). Evaluate for your model and latency/throughput goals.
+Speculative decoding is a technique to reduce latency when max number of concurrency is low. 
+Depending on the methods, the effective concurrency varies, e.g. 16 to 64.
```
```suggestion
Depending on the methods, the effective concurrency varies, for example, from 16 to 64.
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+      --speculative_config '{"method": "eagle3", "model": "yuhuili/EAGLE3-LLaMA3.1-Instruct-8B", "num_speculative_tokens": 2, "draft_tensor_parallel_size": 1, "dtype": "float16"}' \
+      --port 8001
+
+**Important Notes:**
```
```suggestion
**Important notes**:
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+
+**Important Notes:**
+
+We have observed that more `num_speculative_tokens` causes less acceptance rate of draft model tokens, and loss in throughput. Suggest to set the `num_speculative_tokens` <= 2. 
```
```suggestion
We have observed that more ``num_speculative_tokens`` causes less acceptance rate of draft model tokens and a decline in throughput. As a workaround, set `num_speculative_tokens` to <= 2. 
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+Multi-node checklist & troubleshooting
+--------------------------------------
+
+1. Use ``--distributed-executor-backend ray`` across nodes to manage HIP-visible ranks and RCCL communicators. (Ray is the default for multi-node; explicitly setting this flag is optional.)
```
```suggestion
1. Use ``--distributed-executor-backend ray`` across nodes to manage HIP-visible ranks and RCCL communicators. (``ray`` is the default for multi-node. Explicitly setting this flag is optional.)
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+We have observed that more `num_speculative_tokens` causes less acceptance rate of draft model tokens, and loss in throughput. Suggest to set the `num_speculative_tokens` <= 2. 
+
+
+Multi-node checklist & troubleshooting
```
```suggestion
Multi-node checklist and troubleshooting
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+
+1. Use ``--distributed-executor-backend ray`` across nodes to manage HIP-visible ranks and RCCL communicators. (Ray is the default for multi-node; explicitly setting this flag is optional.)
+2. Ensure ``/dev/shm`` is shared across ranks (Docker ``--shm-size``, Kubernetes ``emptyDir``), as RCCL uses shared memory for rendezvous.
+3. For GPUDirect RDMA, set ``RCCL_NET_GDR_LEVEL=2`` and verify links (``ibstat``). Requires supported NICs (e.g., ConnectX‑6+).
```
```suggestion
3. For GPUDirect RDMA, set ``RCCL_NET_GDR_LEVEL=2`` and verify links (``ibstat``). Requires supported NICs (for example, ConnectX‑6+).
```

amd-jnovotny commented on docs/how-to/rocm-for-ai/inference-optimization/workload.rst:
```diff
@@ -333,375 +326,1108 @@ hardware counters are also included.
+   See also:
 
+   * vLLM how-to and performance: https://rocm.docs.amd.com/en/latest/how-to/rocm-for-ai/inference-optimization/workload.html#vllm-performance-optimization
+   * vLLM benchmark Docker (with ROCm, PyTorch, vLLM): https://rocm.docs.amd.com/en/latest/how-to/rocm-for-ai/inference/benchmark-docker/vllm.html
```
Do you want bar links here? Recommend standard format

amd-jnovotny reviewed (dismissed):
Looks like a lot, but it's mainly a few items in several places.

peterjunpark committed:
move vllm perf optimization guide to separate page

fix links

minor rst formatting

formatting

peterjunpark commented:
Split off into separate page because it's massive
